src/main/java/com/conveyal/datatools/manager/controllers/api/ServerController.java

```diff
@@ -148,10 +160,41 @@ private static void validateFields(Request req, Document serverDocument) throws
                 AmazonElasticLoadBalancing elb = AmazonElasticLoadBalancingClient.builder().build();
                 List<TargetGroup> targetGroups = elb.describeTargetGroups(describeTargetGroupsRequest).getTargetGroups();
                 if (targetGroups.size() == 0) {
-                    logMessageAndHalt(req, 400, "Invalid value for Target Group ARN. Could not locate Target Group.");
+                    logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "Invalid value for Target Group ARN. Could not locate Target Group.");
                 }
             } catch (AmazonElasticLoadBalancingException e) {
-                logMessageAndHalt(req, 400, "Invalid value for Target Group ARN.");
+                logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "Invalid value for Target Group ARN.");
+            }
+            // Validate instance type
+            if (serverDocument.get("instanceType") != null) {
+                try {
+                    InstanceType.fromValue(serverDocument.get("instanceType").toString());
+                } catch (IllegalArgumentException e) {
+                    String message = String.format("Must provide valid instance type (if none provided, defaults to %s).", DEFAULT_INSTANCE_TYPE);
+                    logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, message);
+                }
+            }
+        }
+        // Server must have name.
+        if (serverDocument.get("name") == null || "".equals(serverDocument.get("name").toString())) {
+            logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "Server must have valid name.");
+        }
+        // Server must have an internal URL (for build graph over wire) or an s3 bucket (for auto deploy ec2).
+        if (serverDocument.get("s3Bucket") == null || "".equals(serverDocument.get("s3Bucket").toString())) {
+            if (!serverDocument.containsKey("internalUrl") || ((List) serverDocument.get("internalUrl")).size() == 0) {
+                logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "Server must contain either internal URL(s) or s3 bucket name.");
+            }
+        } else {
+            // Verify that application has permission to write to/delete from S3 bucket.
+            String key = UUID.randomUUID().toString();
+            String bucket = serverDocument.get("s3Bucket").toString();
+            try {
+                FeedStore.s3Client.putObject(bucket, key, File.createTempFile("test", ".zip"));
+                FeedStore.s3Client.deleteObject(bucket, key);
```

Comment:
Is there an alternative way to check this? This will incur a monetary cost (probably unnoticeable, but more than $0) to put and delete an object.

Reply:
This was the recommended way I found: https://stackoverflow.com/a/17284647/915811

There might be an alternative, but this certainly seems like the most reliable way to handle this. I worry about introducing any complexity to optimize this, but if there's some other simple way, I'm definitely open to it.

Reply:
Ok, cool, let's keep it as is, but please put that stackoverflow link in the code comments.